cmd/sriov-network-config-daemon/service.go

```diff
@@ -102,105 +102,155 @@ func runServiceCmd(cmd *cobra.Command, args []string) error {
 	}
 
 	setupLog.V(2).Info("sriov-config-service", "config", nodeStateSpec)
+
+	supportedNicIds, err := systemd.ReadSriovSupportedNics()
+	if err != nil {
+		return updateSriovResultErr(setupLog, fmt.Errorf("failed to read list of supported nic ids: %v", err))
+	}
+	sriovv1.InitNicIDMapFromList(supportedNicIds)
+
 	hostHelpers, err := helper.NewDefaultHostHelpers()
 	if err != nil {
-		setupLog.Error(err, "failed to create hostHelpers")
-		return updateSriovResultErr("failed to create hostHelpers")
+		return updateSriovResultErr(setupLog, fmt.Errorf("failed to create hostHelpers"))
+	}
+
+	if phase == PhasePre {
+		_, err = hostHelpers.TryEnableRdma()
+		if err != nil {
+			setupLog.Error(err, "warning, failed to enable RDMA")
+		}
+		hostHelpers.TryEnableTun()
+		hostHelpers.TryEnableVhostNet()
+	}
+
+	configPlugin, nodeState, err := getPluginAndState(setupLog, hostHelpers, nodeStateSpec)
+	if err != nil {
+		return updateSriovResultErr(setupLog, err)
+	}
+
+	if configPlugin == nil || nodeState == nil {
+		setupLog.V(0).Info("no plugin for the platform for the current phase", "platform", nodeStateSpec.PlatformType)
+		return updateSriovResultSucceed(setupLog)
 	}
-	platformHelper, err := platforms.NewDefaultPlatformHelper()
+
+	_, _, err = configPlugin.OnNodeStateChange(nodeState)
 	if err != nil {
-		setupLog.Error(err, "failed to create platformHelpers")
-		return updateSriovResultErr("failed to create platformHelpers")
+		return updateSriovResultErr(setupLog,
+			fmt.Errorf("sriov-config-service: failed to run OnNodeStateChange to update the plugin status %v", err))
 	}
 
-	_, err = hostHelpers.TryEnableRdma()
+	err = configPlugin.Apply()
 	if err != nil {
-		setupLog.Error(err, "warning, failed to enable RDMA")
+		err = updateSriovResultErr(setupLog, fmt.Errorf("failed to apply configuration: %v", err))
+	} else {
+		err = updateSriovResultSucceed(setupLog)
 	}
-	hostHelpers.TryEnableTun()
-	hostHelpers.TryEnableVhostNet()
 
-	var configPlugin plugin.VendorPlugin
-	var ifaceStatuses []sriovv1.InterfaceExt
-	if nodeStateSpec.PlatformType == consts.Baremetal {
-		// Bare metal support
-		vars.DevMode = nodeStateSpec.UnsupportedNics
+	setupLog.V(0).Info("shutting down sriov-config-service")
+	return err
+}
+
+func getPluginAndState(setupLog logr.Logger, hostHelpers helper.HostHelpersInterface, conf *systemd.SriovConfig) (
+	plugin.VendorPlugin, *sriovv1.SriovNetworkNodeState, error) {
+	var (
+		configPlugin  plugin.VendorPlugin
+		ifaceStatuses []sriovv1.InterfaceExt
+		err           error
+	)
+	switch conf.PlatformType {
+	case consts.Baremetal:
```

Comment:
should we have a "default" ? just to be on safe side ?

Reply:
If the switch doesn't match, then the `getPluginAndState` function return nil for the configPlugin. We handle this on line 138 and log message that there is no plugin for this platform.